napari/_qt/threads/status_checker.py

```diff
@@ -0,0 +1,53 @@
+from __future__ import annotations
+
+from threading import Event
+from typing import TYPE_CHECKING
+from weakref import ref
+
+from qtpy.QtCore import QObject, QThread, Signal
+
+from napari.utils.notifications import Notification, notification_manager
+
+if TYPE_CHECKING:
+    from napari.components import ViewerModel
+
+
+class StatusChecker(QThread):
+    status_and_tooltip_changed = Signal(object)
+
+    def __init__(self, viewer: ViewerModel, parent: QObject | None = None):
+        super().__init__(parent=parent)
+        self.viewer_ref = ref(viewer)
+        self._event = Event()
+        self._need_status_update = False
+        self._terminate = False
+
+    def trigger_status_update(self) -> None:
+        self._need_status_update = True
+        self._event.set()
+
+    def terminate(self) -> None:
+        self._terminate = True
+        self._event.set()
+
+    def run(self) -> None:
+        self._event.clear()
+        while not self._terminate:
+            if self._need_status_update:
+                self._need_status_update = False
+                self._event.clear()
+                self.calculate_status()
+            else:
+                self._event.wait()
```

Comment:
@Czaki as far as I can tell, `_need_status_update` and `_event` are redundant? (Whenever `_event` is set, `_need_status_update` is set to `True`, and whenever `_event` is cleared, `_need_status_update` is set to `False`?) So, if my understanding is right, I would change the code as follows:

```suggestion
class StatusChecker(QThread):
    """Separate thread dedicated to updating the status bar.
    
    Prior to https://github.com/napari/napari/pull/7146, status bar updates
    happened on the main thread in the viewer model, which could be
    expensive in some layers and resulted in bad performance when some
    layers were selected.
    
    This class puts the job of computing the status into a separate thread,
    which then uses Qt Signals to update the main viewer with the status
    string.
    
    Because the thread runs a single infinite while loop, the updates are
    naturally throttled to as fast as they can be computed, but no faster.
    
    Attributes
    ----------
    _need_status_update : threading.Event
        An Event (fancy thread-safe bool) for keeping track of when the
        status needs updating (because the cursor has moved).
    _terminate : bool
        Whether the thread needs to be terminated. Set by the QtViewer
        when it is being closed. No more status updates will take place if
        _terminate is set.
    viewer_ref : weakref.ref[napari.viewer.ViewerModel]
        A weak reference to the viewer providing the status updates. We
        don't want this thread to prevent the viewer from being garbage
        collected, so we keep only a weak reference, and check it when
        we need to compute a new status update.
    """

    status_and_tooltip_changed = Signal(object)

    def __init__(self, viewer: ViewerModel, parent: QObject | None = None):
        super().__init__(parent=parent)
        self.viewer_ref = ref(viewer)
        self._need_status_update = Event()
        self._need_status_update.clear()
        self._terminate = False

    def trigger_status_update(self) -> None:
        self._need_status_update.set()

    def terminate(self) -> None:
        self._terminate = True
        self._need_status_update.set()

    def run(self) -> None:
        while not self._terminate:
            if self._need_status_update.is_set():
                self._need_status_update.clear()
                self.calculate_status()
            else:
                self._need_status_update.wait()
```

(Whether or not my understanding is correct, some version of the docstring should be added...)

Reply:
Looks like it is obsolete. I try to create deadlock but fail